calico_node/allocateipip/allocate_ipip_addr.go

```diff
@@ -140,33 +140,46 @@ func removeHostTunnelAddr(ctx context.Context, c client.Interface, nodename stri
 	}
 }
 
+// Make best effort to release IPAM assigned IP address.
+// Log, but otherwise ignore errors.
+func releaseIPAddrByHandle(ctx context.Context, c client.Interface, handle string, ipv4Addr string) {
+	err := c.IPAM().ReleaseByHandle(ctx, handle)
+	if err != nil {
+		log.WithError(err).WithField("IP", ipv4Addr).Error("Error releasing IP address on failure")
+	} else {
+		log.WithField("IP", ipv4Addr).Debug("Released IP address")
+	}
+}
+
 // assignHostTunnelAddr claims an IPIP-enabled IP address from the first pool
 // with some space. Stores the result in the host's config as its tunnel
 // address.
 func assignHostTunnelAddr(ctx context.Context, c client.Interface, nodename string, ipipCidrs []net.IPNet) {
+	handle := "ipipAddr:" + nodename
 	args := ipam.AutoAssignArgs{
 		Num4:      1,
 		Num6:      0,
-		HandleID:  nil,
+		HandleID:  &handle,
 		Attrs:     nil,
 		Hostname:  nodename,
 		IPv4Pools: ipipCidrs,
 	}
 
 	ipv4Addrs, _, err := c.IPAM().AutoAssign(ctx, args)
 	if err != nil {
-		log.WithError(err).Fatal("Unable to autoassign an address for IPIP")
+		log.WithError(err).Fatal("Unable to auto-assign an address for IPIP")
 	}
 
 	if len(ipv4Addrs) == 0 {
-		log.Fatal("Unable to autoassign an address for IPIP - pools are likely exhausted.")
+		log.Fatal("Unable to auto-assign an address for IPIP - pools are likely exhausted.")
 	}
 
 	var updateError error
 	// If the update fails with ResourceConflict error then retry 5 times with 1 second delay before failing.
 	for i := 0; i < 5; i++ {
 		node, err := c.Nodes().Get(ctx, nodename, options.GetOptions{})
 		if err != nil {
```

Comment:
This logic looks a bit flawed - if we hit an error shouldn't we be exiting this function rather than trying to carry on blindly?

Reply:
it feels like this method should just return an error and the calling code should do the release if we hit an error in the allocation.

Reply:
Oh actually - my thinking is flawed too.

So this is trying to loop 5 times to retry in the event of a failure.  In that case you should not be releasing the IP address until we've completely given up.

Reply:
I'm not following your comment. We loop through 5 times, unless we hit an error from c.Nodes().Get(), in that case we're calling `Fatalf()` and bailing. I added the `releaseIPAddrByHandle()` right before we exit.